### PR TITLE
Add sample browser app and supporting code

### PR DESCRIPTION
**Pull request by pocmo**
Okay, I took my prototype code and with that created the sample-browser app module as well as the supporting code in browser-toolbar, browser-session and engine-*.

Some highlights:

* The sample app loads an initial page (works with geckoview and webview) and updates the toolbar whenever the user follows a link. This is a bit simpler than my prototype so far - but the code is closer to what we want to do. And as soon as this is in we can add more things in parallel.

* I created a new concept-* group for abstract definitions of modules. browser-engine was such a thing already and it is concept-engine now. Additionally I created concept-toolbar with just interfaces that browser-toolbar implements. The idea is that the other modules only depend on concept-toolbar which allows us to switch the toolbar implementation. An example use case is Firefox for Fire TV: It could use the same engine, session and toolbar supporting code (see below) but the actual toolbar might be a totally different one, because TVs are just different things.

* I created feature-* modules that are combining multiple modules and contain the glue code. As those modules compile against concept-* modules the actual implementation of things can differ. I created a feature-session module which connects browser-session and concept-engine: Whenever the engine state changes the corresponding session object is updated. And I created feature-toolbar: Whenever the session changes the toolbar will be updated and (not part of this code yet) when the user types into the toolbar then the engine will be asked to load the page.

### CODE REVIEW

**Comment by pocmo** [comment]
Oh, and: I decided to create an engine-system implementation (based on WebView). Even though we may not implement it fully (will not be needed for now) it was still helpful to make sure the interfaces in concept-engine make sense and are not crafted to only support GeckoView.

File: components/concept/toolbar/src/main/java/mozilla/components/definition/toolbar/ToolbarConcept.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.definition.toolbar
```
Comment (csadilek) [inline components/concept/toolbar/src/main/java/mozilla/components/definition/toolbar/ToolbarConcept.kt]: We introduce .definition. here (in addition to concept) which concept-engine doesn't have. My personal preference would be to keep the namespace simple i.e. stick to mozilla.components.toolbar and just Toolbar as the classname. Basically, the same as you did with Engine.

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarFeature.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+package mozilla.components.feature.toolbar
+
+import mozilla.components.definition.toolbar.ToolbarConcept
+import mozilla.components.session.SessionManager
+
+/**
+ * Feature implementation for connecting a toolbar implementation with the session module.
+ */
+class ToolbarFeature(
```
Comment (csadilek) [inline components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarFeature.kt]: We should probably introduce a common interface `Feature` with a `start` and `stop` method but we can do that later as then we also need a "parent" feature module to host that interface.

File: components/browser/engine-system/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<!-- This Source Code Form is subject to the terms of the Mozilla Public
+   - License, v. 2.0. If a copy of the MPL was not distributed with this
+   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="org.mozilla.engine.system" />
```
[the remark this answers is not in the document]
Comment (pocmo) [reply components/browser/engine-system/src/main/AndroidManifest.xml]: 👍 Yeah, I think we should update all the manifests. I'll file an issue for that: #60

File: components/browser/session/src/main/java/mozilla/components/session/SessionManager.kt [hunk 4]
```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.session
+
+/**
+ * This class provides access to a centralized registry of all active sessions.
+ */
```
[the remark this answers is not in the document]
Comment (pocmo) [reply components/browser/session/src/main/java/mozilla/components/session/SessionManager.kt]: I'd leave that open. Technically there's no limitation to one - although we probably won't need multiple for now. I guess you could build something like "multiple windows" on desktop with multiple session managers.

File: components/browser/session/src/main/java/mozilla/components/session/Session.kt [hunk 5]
```diff
@@ -4,4 +4,48 @@
...
+    initialUrl: String
+) {
+    /**
+     * Interface to be implemented by classes that want to observe a session.
+     */
+    interface Observer {
+        fun onUrlChanged()
+    }
+
+    private val observers = mutableListOf<Observer>()
```
[the remark this answers is not in the document]
Comment (pocmo) [reply components/browser/session/src/main/java/mozilla/components/session/Session.kt]: 👍

File: components/browser/session/src/main/java/mozilla/components/session/SessionManager.kt [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+        get() = sessions.size
+
+    /**
+     * Get the currently selected session. Only one session can be selected.
+     */
+    val selectedSession: Session
+        get() = sessions.selected
+
+    /**
+     * Register an observer to get notified about changes regarding sessions (e.g. adding a new
```
[the remark this answers is not in the document]
Comment (pocmo) [reply components/browser/session/src/main/java/mozilla/components/session/SessionManager.kt]: Makes sense to be consistent 👍

File: components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarFeature.kt (hunk 2 above)
Comment (pocmo) [reply components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarFeature.kt]: Yeah, I was thinking that too (and I had that in my prototype). I wasn't sure if this was just a thing required by those two specifically or if this is a pattern that will be followed by all features more or less.

concept-feature? :)

File: components/browser/engine-system/src/main/AndroidManifest.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (pocmo) [reply components/browser/engine-system/src/main/AndroidManifest.xml]: I'll add a commit for this on top of the PR.

[the remark this answers is not in the document]
Comment (pocmo) [reply components/browser/engine-system/src/main/AndroidManifest.xml]: Nope. Just removed the commit again. If we want to match the package names then we need to make sure that the packages are unique. It doesn't matter in the Java world but it matters when Android tries to merge resources and generated code - You can't have two modules with the same Android package or everything explodes. Let's figure that out in #60.

**Comment by pocmo** [comment]
Okay, I think I added commits addressing all you comments.

File: components/browser/engine-system/src/main/java/mozilla/components/engine/gecko/SystemEngine.kt [hunk 7]
```diff
@@ -0,0 +1,30 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.engine.gecko
```
Comment (csadilek) [inline components/browser/engine-system/src/main/java/mozilla/components/engine/gecko/SystemEngine.kt]: This should be engine.system? Sorry, might have missed that before.

File: components/feature/toolbar/src/main/java/mozilla/components/feature/session/ToolbarPresenter.kt [hunk 8]
```diff
@@ -0,0 +1,60 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.session
```
Comment (csadilek) [inline components/feature/toolbar/src/main/java/mozilla/components/feature/session/ToolbarPresenter.kt]: This should be feature.toolbar.

File: components/browser/engine-system/src/main/java/mozilla/components/engine/gecko/SystemEngine.kt (hunk 7 above)
Comment (pocmo) [reply components/browser/engine-system/src/main/java/mozilla/components/engine/gecko/SystemEngine.kt]: Argh. I think what happened is that Android Studio didn't only rename the package in the module I'm currently in but also renames it in all other modules. That's unexpected.

**Comment by pocmo** [comment]
Okay, fixed the package names. :)

**Comment by csadilek** [comment]
Great, ready to go. Do you want to squash the commits or keep separately?

**Comment by pocmo** [comment]
I'll keep them. :)

**Pull request merged**